drivers/mpl3115a2/mpl3115a2.c

```diff
@@ -16,211 +17,180 @@
+#include <string.h>
+
+#include "log.h"
 #include "periph/i2c.h"
+
 #include "mpl3115a2.h"
 #include "mpl3115a2_reg.h"
 
 #define ENABLE_DEBUG    (0)
 #include "debug.h"
 
-#define I2C_SPEED                  I2C_SPEED_FAST
+#define I2C_SPEED           I2C_SPEED_FAST
 
-int mpl3115a2_test(mpl3115a2_t *dev)
-{
-    uint8_t reg;
+#define BUS                 (dev->params.i2c)
+#define ADDR                (dev->params.addr)
 
-    /* Acquire exclusive access to the bus. */
-    i2c_acquire(dev->i2c);
-    if (i2c_read_regs(dev->i2c, dev->addr, MPL3115A2_WHO_AM_I, &reg, 1) != 1) {
-        /* Release the bus for other threads. */
-        i2c_release(dev->i2c);
-        return -1;
-    }
-
-    i2c_release(dev->i2c);
-    if (reg != MPL3115A2_ID) {
-        return -1;
-    }
-
-    return 0;
-}
-
-int mpl3115a2_init(mpl3115a2_t *dev, i2c_t i2c, uint8_t address, uint8_t os_ratio)
+int mpl3115a2_init(mpl3115a2_t *dev, const mpl3115a2_params_t *params)
 {
     uint8_t reg;
 
-    /* write device descriptor */
-    dev->i2c = i2c;
-    dev->addr = address;
-    dev->initialized = false;
+    assert(dev);
+    assert(params);
 
-    if (os_ratio > MPL3115A2_OS_RATIO_128) {
-        return -1;
-    }
+    /* write device descriptor */
+    memcpy(dev, params, sizeof(mpl3115a2_params_t));
 
-    i2c_acquire(dev->i2c);
+    i2c_acquire(BUS);
     /* initialize the I2C bus */
-    if (i2c_init_master(i2c, I2C_SPEED) < 0) {
-        i2c_release(dev->i2c);
-        return -2;
+    if (i2c_init_master(BUS, I2C_SPEED) < 0) {
+        i2c_release(BUS);
+        LOG_ERROR("mpl3115a2_init: failed to init I2C!\n");
+        return -MPL3115A2_ERROR_I2C;
     }
-    i2c_release(dev->i2c);
-
-    if (mpl3115a2_test(dev)) {
-        return -3;
+    /* test device */
+    if (i2c_read_regs(BUS, ADDR, MPL3115A2_WHO_AM_I, &reg, 1) != 1) {
+        /* Release the bus for other threads. */
+        i2c_release(BUS);
+        LOG_ERROR("mpl3115a2_init: I2C error!\n");
+        return -MPL3115A2_ERROR_I2C;
     }
-
-    reg = MPL3115A2_CTRL_REG1_OS(os_ratio);
-
-    i2c_acquire(dev->i2c);
-    if (i2c_write_regs(dev->i2c, dev->addr, MPL3115A2_CTRL_REG1, &reg, 1) != 1) {
-        i2c_release(dev->i2c);
-        return -4;
+    if (reg != MPL3115A2_ID) {
+        LOG_ERROR("mpl3115a2_init: invalid WHO_AM_I value (0x%02x)!\n", (int)reg);
+        return -MPL3115A2_ERROR_DEV;
     }
-    i2c_release(dev->i2c);
-
-    reg = MPL3115A2_PT_DATA_CFG_TDEFE
-          | MPL3115A2_PT_DATA_CFG_PDEFE
-          | MPL3115A2_PT_DATA_CFG_DREM;
-
-    i2c_acquire(dev->i2c);
-    if (i2c_write_regs(dev->i2c, dev->addr, MPL3115A2_PT_DATA_CFG, &reg, 1) != 1) {
-        i2c_release(dev->i2c);
-        return -4;
+    /* set sample rate */
+    reg = MPL3115A2_CTRL_REG1_OS(dev->params.rate);
+    if (i2c_write_regs(BUS, ADDR, MPL3115A2_CTRL_REG1, &reg, 1) != 1) {
+        i2c_release(BUS);
+        LOG_ERROR("mpl3115a2_init: failed to set sample rate!\n");
+        return -MPL3115A2_ERROR_CNF;
```

Comment:
If the I2C read error code is kept, this one could be renamed to MPL3115A2_ERROR_WRITE. What do you think ?